### ISSUE TITLE
False positive: Format Validation for ruby built-in constant regular expression

### ISSUE THREAD

**Issue by nlamantia** [presidentbeef/brakeman#1632]
### Background

Brakeman version: 5.1.1
Rails version: 6.0.3.7
Ruby version: 2.7.4

Link to Rails application code: ?
#### False Positive

```
Confidence: High
--
  | [2021-08-11T21:21:06Z] Category: Format Validation
  | [2021-08-11T21:21:06Z] Check: ValidationRegex
  | [2021-08-11T21:21:06Z] Message: Insufficient validation for `initiated_by` using `/[a-zA-Z0-9.!\#$%&'*+\/=?^_`{\|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)/`. Use `\A` and `\z` as anchors
```

Relevant code:

```ruby
validates :initiated_by, :format => { :with => URI::MailTo::EMAIL_REGEXP }
```

_Why might this be a false positive?_

`URI::MailTo::EMAIL_REGEXP` evaluates to ```/\A[a-zA-Z0-9.!\#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*\z/```. So the flagged code does use a regex that is properly anchored, but it keeps popping up as a warning. If I create my own constant for the exact same regex, as below, it is not flagged as a security warning:

```ruby
EMAIL_REGEX = /\A[a-zA-Z0-9.!\#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*\z/
...
validates :intiated_by, :format => { :with => EMAIL_REGEX }
```

**Comment by presidentbeef** [comment]
Hi @naokikimura.

The regex and code you are referencing do not match the warning.

Note the warning is about the `initiated_by` attribute and the regex does not have `\A` or `\z` anchors.

Brakeman doesn't know what `URI::MailTo::EMAIL_REGEXP` is so the regex in the warning must exist in your codebase.

**Comment by nlamantia** [comment]
Hey @presidentbeef,

Sorry about that. Copy and paste error. `initiated_by` is the same type of field as `email` just on a different model. My bad there. I corrected the comment above.

As for `URI::MailTo::EMAIL_REGEXP`, I feel that we shouldn't have to make a direct copy of constants that are included with Ruby as it violates "DRY". What's the rationale for this behavior?

**Comment by presidentbeef** [comment]
I'm not telling you to copy the constant.

My point is that 

    /[a-zA-Z0-9.!\#$%&'*+\/=?^_`{\|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)/

must literally exist in your code somewhere. It is not the same as the regexp value of  Otherwise Brakeman wouldn't warn about it, because it wouldn't know about it.

It's not the same as the value of `URI::MailTo::EMAIL_REGEXP` that you have reported here.

**Comment by nlamantia** [comment]
That regular expression is nowhere in the code base. The anchored version of that is `URI::MailTo::EMAIL_REGEXP`, but I searched the entire code base for that exact regular expression and it's not in there. Even if it was somewhere in the code base, the error is pointing to the line that uses `URI::MailTo::EMAIL_REGEXP`.

**Comment by nlamantia** [comment]
@presidentbeef So after looking deeper into this, we found a similar regular expression with the same name. _However_, it the similar regular expression was not in the same namespace as the one we were referencing in our code. Specifically, the code in question was
```ruby
validates :initiated_by, :format => { :with => URI::MailTo::EMAIL_REGEXP }
```
We have a constant elsewhere in our code (aka in a different namespace altogether) as so
```ruby
class EmailMessageService
    ...
    EMAIL_REGEX = %r{[a-zA-Z0-9.!\#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)}.freeze # Relevent part pulled from URI::MailTo
   ...
end
```
If we wanted to refer to this constant instead of the one we are currently referring to, since it's in a different namespace, we would have to change the code to 
```ruby
validates :initiated_by, :format => { :with => EmailMessageService::EMAIL_REGEXP }
```
So the warning is picking up a constant named `EMAIL_REGEXP`, but it's in a different namespace than the one we are actually trying to validate against. We got around it by changing the name of that constant, but we still don't think this behavior is intended.
